### PR TITLE
Update pac4j to 5.2.0

### PR DESCRIPTION
**Pull request by Riliane**
See [JENKINS-69462](https://issues.jenkins-ci.org/browse/JENKINS-69462).

This needs https://github.com/jenkins-infra/helpdesk/issues/3108 to successfully be built - now it will fail because of the shibboleth artifacts.
Now that we are dropping Java 8 we can update pac4j further.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Appropriate autotests or explanation to why this change has no tests

### CODE REVIEW

**Comment by kuisathaverat** [comment]
I am preparing the plugin for this update, first I have to merge https://github.com/jenkinsci/saml-plugin/pull/261

File: pom.xml [hunk 1]
```diff
@@ -196,6 +195,12 @@ under the License.
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+    <groupId>org.pac4j</groupId>
+    <artifactId>pac4j-jee</artifactId>
+    <version>5.2.0</version>
+    <optional>false</optional>
+    </dependency>
```
Comment (kuisathaverat) [inline pom.xml:203]:
```suggestion
    <dependency>
        <groupId>org.pac4j</groupId>
        <artifactId>pac4j-jee</artifactId>
        <version>5.2.0</version>
        <optional>false</optional>
    </dependency>
```

**Comment by kuisathaverat** [comment]
I have resolve the conflicts and update the major version

File: pom.xml [hunk 2]
```diff
@@ -196,6 +195,12 @@ under the License.
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+        <groupId>org.pac4j</groupId>
+        <artifactId>pac4j-jee</artifactId>
+        <version>5.2.0</version>
+        <optional>false</optional>
+    </dependency>
```
Comment (Vlatombe) [inline pom.xml:203]:
```suggestion
    <dependency>
      <groupId>org.pac4j</groupId>
      <artifactId>pac4j-jee</artifactId>
      <version>5.2.0</version>
    </dependency>
```

Comment (Vlatombe) [reply pom.xml:203]: I assume this is because `org.pac4j.core.context.JEEContext` has been moved to this module?

**Comment by kuisathaverat** [comment]
I will try to make some manual testing, If I do not see anything weird, I would merge and release a new version.

File: pom.xml (hunk 2 above)
Comment (Riliane) [reply pom.xml:203]: yes, it is. Former `J2EContext` is now `JEEContext` and in this module

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 3]
```diff
@@ -255,15 +257,15 @@ public HttpResponse doCommenceLogin(final StaplerRequest request, final StaplerR
...
 
-        RedirectAction action = new SamlRedirectActionWrapper(getSamlPluginConfig(), request, response).get();
-        if (action.getType() == RedirectType.REDIRECT) {
-            LOG.fine("REDIRECT : " + action.getLocation());
-            return HttpResponses.redirectTo(action.getLocation());
-        } else if (action.getType() == RedirectType.SUCCESS) {
-            LOG.fine("SUCCESS : " + action.getContent());
-            return HttpResponses.literalHtml(action.getContent());
+        RedirectionAction action = new SamlRedirectActionWrapper(getSamlPluginConfig(), request, response).get();
+        if (action instanceof SeeOtherAction) {
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]: Redirections do not work, if the server returns a 302 code this alternative is not evaluated, I guess is not the correct type for redirections.

Comment (kuisathaverat) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303

Comment (kuisathaverat) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302

Comment (kuisathaverat) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]: FoundAction must be added

Comment (kuisathaverat) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]: https://moz.com/learn/seo/redirection, it must manage 301, 302, and 307 codes

Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:261]:
```suggestion
        if (action instanceof SeeOtherAction || action instanceof FoundAction) {
```

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 4]
```diff
@@ -255,15 +257,15 @@ public HttpResponse doCommenceLogin(final StaplerRequest request, final StaplerR
...
+            LOG.fine("REDIRECT : " + ((SeeOtherAction)action).getLocation());
+            return HttpResponses.redirectTo(((SeeOtherAction)action).getLocation());
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:263]:
```suggestion
            LOG.fine("REDIRECT : " + ((WithLocationAction)action).getLocation());
            return HttpResponses.redirectTo(((WithLocationAction)action).getLocation());
```

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 5]
```diff
@@ -45,8 +45,10 @@
 import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.StaplerResponse;
 import org.kohsuke.stapler.interceptor.RequirePOST;
-import org.pac4j.core.redirect.RedirectAction;
-import org.pac4j.core.redirect.RedirectAction.RedirectType;
+import org.pac4j.core.exception.http.OkAction;
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:48]:
```suggestion
import org.pac4j.core.exception.http.FoundAction;
import org.pac4j.core.exception.http.WithLocationAction;
import org.pac4j.core.exception.http.OkAction;
```

File: pom.xml [hunk 6]
```diff
@@ -98,8 +98,7 @@ under the License.
     <dependency>
       <groupId>org.pac4j</groupId>
       <artifactId>pac4j-saml</artifactId>
-      <!-- versions 4.x.x require JDK 11 -->
-      <version>3.9.0</version>
+      <version>5.2.0</version>
```
Comment (kuisathaverat) [inline pom.xml:101]: Why not the latest version `5.4.6`?

```suggestion
      <version>5.4.6</version>
```

File: pom.xml [hunk 7]
```diff
@@ -196,6 +195,11 @@ under the License.
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.pac4j</groupId>
+      <artifactId>pac4j-jee</artifactId>
+      <version>5.2.0</version>
```
Comment (kuisathaverat) [inline pom.xml:201]:
```suggestion
      <version>5.4.6</version>
```

Comment (kuisathaverat) [reply pom.xml:201]: JEESessionStore is deprecated in 5.4.6, I am checking why.

Comment (kuisathaverat) [reply pom.xml:201]: >Use the pac4j-javaee dependency instead.

File: pom.xml [hunk 8]
```diff
@@ -196,6 +195,11 @@ under the License.
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.pac4j</groupId>
+      <artifactId>pac4j-jee</artifactId>
```
Comment (kuisathaverat) [inline pom.xml:200]:
```suggestion
      <artifactId>pac4j-javaee</artifactId>
```

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 9]
```diff
@@ -18,14 +18,16 @@
...
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.logging.Logger;
 import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.StaplerResponse;
 import org.opensaml.core.config.InitializationException;
 import org.opensaml.core.config.InitializationService;
-import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.context.JEEContext;
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java:28]:
```suggestion
import org.pac4j.jee.context.JEEContext;
```

File: src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java [hunk 10]
```diff
@@ -21,7 +21,8 @@
 import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.StaplerResponse;
 import org.pac4j.core.context.WebContext;
-import org.pac4j.core.exception.HttpAction;
+import org.pac4j.core.context.session.JEESessionStore;
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java:24]:
```suggestion
import org.pac4j.jee.context.JEEContext;;
```

File: src/main/java/org/jenkinsci/plugins/saml/SamlRedirectActionWrapper.java [hunk 11]
```diff
@@ -19,15 +19,16 @@
 
 import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.StaplerResponse;
+import org.pac4j.core.context.session.JEESessionStore;
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlRedirectActionWrapper.java:22]:
```suggestion
import org.pac4j.jee.context.JEEContext;
```

**Comment by kuisathaverat** [comment]
Summary of my tests, there is an error in the redirections 302 code is not correctly processed. The version of pac4j used is not the latest, is from about a year ago, it is possible to update to the latest with minor changes.

I have add comments with the changes needed to fix the 302 redirections and bump the pac4j version to 5.4.6

**Comment by Riliane** [comment]
committed suggestions to pull them but still got things to fix (JEESessionStore)

**Review by kuisathaverat — APPROVED** [review]
tested Manually

**Comment by jkbszpg** [comment]
the new pac4j-saml 5.4.6, requires an Optional "Destination" field in the SAML response to be present unless disabled with a setting
```
cfg.setResponseDestinationAttributeMandatory(false);
```
https://github.com/pac4j/pac4j/pull/1871

resulting in
```
WARNING o.j.p.saml.SamlSecurityRealm#doFinishLogin: Unable to validate the SAML Response: SAML configuration does not allow response Destination to be null
For more info check 'Maximum Authentication Lifetime' at https://github.com/jenkinsci/saml-plugin/blob/master/doc/CONFIGURE.md#configuring-plugin-settings
If you have issues check the troubleshoting guide at https://github.com/jenkinsci/saml-plugin/blob/master/doc/TROUBLESHOOTING.md
org.pac4j.saml.exceptions.SAMLEndpointMismatchException: SAML configuration does not allow response Destination to be null
        at org.pac4j.saml.profile.impl.AbstractSAML2ResponseValidator.verifyEndpoint(AbstractSAML2ResponseValidator.java:204)
        at org.pac4j.saml.sso.impl.SAML2AuthnResponseValidator.validateSamlProtocolResponse(SAML2AuthnResponseValidator.java:247)
        at org.pac4j.saml.sso.impl.SAML2AuthnResponseValidator.validate(SAML2AuthnResponseValidator.java:91)
        at org.pac4j.saml.profile.impl.AbstractSAML2MessageReceiver.receiveMessage(AbstractSAML2MessageReceiver.java:53)
        at org.pac4j.saml.sso.impl.SAML2WebSSOProfileHandler.receive(SAML2WebSSOProfileHandler.java:35)
        at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.receiveLogin(SAML2CredentialsExtractor.java:71)
        at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.extract(SAML2CredentialsExtractor.java:66)
        at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:71)
        at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:145)
        at org.jenkinsci.plugins.saml.SamlProfileWrapper.process(SamlProfileWrapper.java:56)
Caused: org.springframework.security.authentication.BadCredentialsException: SAML configuration does not allow response Destination to be null
        at org.jenkinsci.plugins.saml.SamlProfileWrapper.process(SamlProfileWrapper.java:61)
        at org.jenkinsci.plugins.saml.SamlProfileWrapper.process(SamlProfileWrapper.java:35)
```
is there a way for saml-plugin to pass this config through?

**Comment by kuisathaverat** [comment]
@jkbszpg @rdpa-bah I am not subscribed to PR (so comments here would not arrive in my inbox), I will take a look, it is simple to add that option.

**Pull request merged**